core/src/main/java/org/apache/iceberg/data/avro/IcebergDecoder.java

```diff
@@ -106,7 +100,10 @@ public void addSchema(org.apache.iceberg.Schema writeSchema) {
 
   private void addSchema(Schema writeSchema) {
     long fp = SchemaNormalization.parsingFingerprint64(writeSchema);
-    decoders.put(fp, new RawDecoder<>(readSchema, writeSchema));
+    decoders.put(
+        fp,
+        new RawDecoder<>(
+            readSchema, avroSchema -> DataReader.create(readSchema, avroSchema), writeSchema));
```

Comment:
The formatting is awkward because there is too much on one line. Can you clean this up a little by moving the lambda or creating the decoder separately?